compiler/backend_inkwell/src/lib.rs

```diff
@@ -0,0 +1,1041 @@
+                        .iter()
+                        .map(|arg| self.get_value_with_id(function_ctx, arg).unwrap().into())
+                        .collect();
+                    if let Some(FunctionInfo {
+                        function_value,
+                        captured_ids: _,
+                        env_type,
+                    }) = self.functions.get(function)
+                    {
+                        if env_type.is_some() {
+                            let get_candy_fn_env = self
+                                .module
+                                .get_function("get_candy_function_environment")
+                                .unwrap();
+
+                            let fn_object = self.values.get(function).unwrap_or_else(|| {
+                                panic!("Function {function} should have global visibility")
+                            });
+
+                            let fn_env_ptr = self.builder.build_call(
+                                get_candy_fn_env,
+                                &[fn_object.as_pointer_value().into()],
+                                "",
+                            );
+
+                            args.push(fn_env_ptr.try_as_basic_value().unwrap_left().into());
+                        }
+                        let call = self.builder.build_call(**function_value, &args, "");
+                        let call_value = Rc::new(call.try_as_basic_value().unwrap_left());
+                        self.locals.insert(*id, call_value.clone());
+
+                        if idx == mir.expressions.len() - 1 {
+                            self.builder
+                                .build_return(Some(&call_value.into_pointer_value()));
+                        }
+                    } else {
+                        let function_value = self
+                            .get_value_with_id(function_ctx, function)
+                            .unwrap_or_else(|| panic!("There is no function with ID {function}"));
+
+                        let get_candy_fn_ptr = self
+                            .module
+                            .get_function("get_candy_function_pointer")
+                            .unwrap();
+                        let get_candy_fn_env = self
+                            .module
+                            .get_function("get_candy_function_environment")
+                            .unwrap();
+
+                        let fn_ptr =
+                            self.builder
+                                .build_call(get_candy_fn_ptr, &[function_value.into()], "");
+
+                        let fn_env_ptr =
+                            self.builder
+                                .build_call(get_candy_fn_env, &[function_value.into()], "");
+
+                        args.push(fn_env_ptr.try_as_basic_value().unwrap_left().into());
+
+                        let candy_fn_type = candy_type_ptr.fn_type(&[], true);
```

Comment:
Since this is the path for an indirect function call, we don't know how many parameters the function has (since functions with different signatures are saved as the same type). It might be possible to save that information in the function object, but extracting it would incur further FFI costs and the compiler should have verified at an earlier stage that the number of parameters is correct.